### ISSUE TITLE
Using a byref parameter with SRTP results in code which always fails

### ISSUE THREAD

**Issue by BillHally** [fable-compiler/Fable#3328]
### Description

Using a `byref` parameter with SRTP results in code which always fails, because of 2 problems with the code generation.

### Repro code

```fsharp
type TypeWithByRefMember() =
  static member DoubleIntByRef (x: byref<int>) : unit = x <- x + x

let inline doubleIntByRef (x: ^a) (input: int) : int =
    let mutable value = input
    (^a: (static member DoubleIntByRef: byref<int> -> unit) &value)
    value

try
    let result = doubleIntByRef (TypeWithByRefMember()) 7
    printfn $"result: {result}"
with
| ex -> eprintfn $"ERROR: {ex}"
```

View in the [repl](https://fable.io/repl/#?code=PYBwpgdgBAygngZwC5gLYFgBQXSSgMQEMAjAGzB3GiLLADoBhYAJws12pPMZfoCkEASQgpmoLFiRxwUACrSwAdQCWSABYAhOACUwAMwCyaYmGYAKAJRQAvFihRkhJMoDGUVMdNQAIsACutMJIWrp6UGYAHgBcUMRwrHoAPMoiAHxWMX4QqjZQEVCJALR5UADUeRKY5EhQKaQpYFAAJv6BIiH64dFQAHqEVmYpIH5IMSlIGbUiNnb2UNXuI1yNAG6EpH6N1lPDSLP2Zn0xZo7Obh6oJsw+reRBHXoxcQnJaVCFqVBZqlYAZGsbMAWfZQAGbSpIZhwEELVgIPykGrbFoBO7tHSdMzycAqdQPIyXUyWKwAdhBIGY4z00AAJAAiOEI0ZQADejMRAF86VgAO6qNRYAA+UDA+Q+IopVNpdIAotptAB5bQxFmirlYIA&html=Q&css=Q).

##### Javascript

Skipping the `TypeWithByRefMember` code, I think the following would be the expected code (at least, cutting and pasting it into the repl gives the expected output):

```javascript
try {
    let result;
    const x = TypeWithByRefMember_$ctor();
    let value = 7;
    const arg0_ = (new FSharpRef(() => value, (v) => {
        value = (v | 0);
    }));
    TypeWithByRefMember_DoubleIntByRef_Z3D2991E8(arg0_);
    result = value;
    toConsole(`result: ${result}`);
}
catch (ex) {
    toConsoleError(`ERROR: ${ex}`);
}
```

##### Output

```
result: 14
```

##### Javascript

Again, skipping the `TypeWithByRefMember` code:

```javascript
try {
    let result;
    const x = TypeWithByRefMember_$ctor();
    let value = 7;
    const arg0_ = (new FSharpRef(() => value, (v) => {
        value = (v | 0);
    })) | 0; // <-- 1. Performs bitwise OR with 0
    TypeWithByRefMember_DoubleIntByRef_Z3D2991E8(arg0_.contents); // <-- 2. Passes arg0_.contents instead of args0_
    result = value;
    toConsole(`result: ${result}`);
}
catch (ex) {
    toConsoleError(`ERROR: ${ex}`);
}
```

So it looks like 2 issues with the Javascript code generation:
1. performing a bitwise OR with 0 on the `FSharpRef` object (so `arg0_` is always set to `0`)
2. even if the 1st issue was fixed (i.e. the bitwise OR was removed), the `FSharpRef` object's `contents` property is passed instead of the object itself

##### Output

```
ERROR: TypeError: Cannot read properties of undefined (reading 'contents')
```

### Related information

* Fable version:
  * 4.0.0-theta-018 (project where I noticed the issue)
  * 4.0.0-theta-008 (repl where the issue can be replicated)

* Operating system
  * Windows (project where I noticed the issue)
  * Whatever the repl runs on (repl where the issue can be replicated)

**Comment by weebs** [comment]
I'm glad you posted this, I was playing with byref's recently and noticed a byref's contents getting passed to other functions instead of the byref itself when calling certain things. I'm not that savvy with F#'s different ref types so I didn't think much of it at first

I had been digging around the byref replacement code recently and I noticed the byref in your code getting replaced in a familiar spot and I think this might be the crux of the issue:

```diff
             let v = makeValueFrom com ctx r var
-            if isByRefValue var && com.Options.Language <> Rust then
+            if isByRefValue var &&
+               // The replacement only needs to happen when var.FullType = byref<fsExpr.Type>
+               fsExpr.Type = var.FullType.GenericArguments.[0] &&
+               com.Options.Language <> Rust then
```

This replacement of `a_byref` to `a_byref.contents` needs to happen when a byref is the last expression in something that returns its underlying value type, but it was also getting replaced in situations like passing a byref as a function argument. In this case it happens because the inlined function passes a byref value. In my case I had a function that took a byref and called another function with the byref

I'm not sure if this is the proper way to fix the issue but it seems to work in the test cases I've tried. I was having trouble getting the tests build target to run on my machine so I couldn't verify all the other tests still worked but I've pushed my commit to a fork in hopes it helps. I've also included a second commit I'm not as sure about, but I noticed during testing that the JS output would do `(new FSharpRef...) | 0` on const byref variables (which I was only able to trigger by the below code)

```fsharp
let byrefFunc(n: byref<int>) =
    n <- n + n

let inline callWithByrefCreatedFromByrefInlined(n: byref<int>) =
    let f = &n
    byrefFunc &f

let mutable an_int = 22
callWithByrefCreatedFromByrefInlined &an_int
```

byref.contents replacement:
https://github.com/weebs/Fable/commit/a237c041c88ac9e00a25bd566cf781b43f7ac0c3

checking for byref in babel wrapIntExpression:
https://github.com/fable-compiler/Fable/commit/3c6321b88def1ce087a03f6ffa7e60bc085c5fb5

**Comment by BillHally** [comment]
@weebs thanks for this - I'll give your changes a try.

**Comment by weebs** [comment]
Good to hear, thanks for the additional testing @BillHally ! I'll submit a PR when I get a chance later today, this will be my first time submitting a PR to a project and I'm excited to knock that off the bucket list 😁 I've been hacking on Fable the past couple weeks and started watching the repo the day before you posted, really quite a coincidence that the first issue I saw was something I knew enough to investigate

**Comment by BillHally** [comment]
> I'll submit a PR when I get a chance later today

That's great @weebs!

**Comment by ncave** [comment]
Fixed by #3329 / #3331.

**Comment by BillHally** [comment]
> Fixed by #3329 / #3331.

Thanks @ncave, and @weebs!
